bridge/gitlab/iterator.go

```diff
@@ -0,0 +1,253 @@
+package gitlab
+
+import (
+	"time"
+
+	"github.com/xanzy/go-gitlab"
+)
+
+type issueIterator struct {
+	page  int
+	index int
+	cache []*gitlab.Issue
+}
+
+type noteIterator struct {
+	page  int
+	index int
+	cache []*gitlab.Note
+}
+
+type labelEventIterator struct {
+	page  int
+	index int
+	cache []*gitlab.LabelEvent
+}
+
+type iterator struct {
+	// gitlab api v4 client
+	gc *gitlab.Client
+
+	// if since is given the iterator will query only the updated
+	// issues after this date
+	since time.Time
+
+	// project id
+	project string
+
+	// number of issues and notes to query at once
+	capacity int
+
+	// sticky error
+	err error
+
+	// issues iterator
+	issue *issueIterator
+
+	// notes iterator
+	note *noteIterator
+
+	// labelEvent iterator
+	labelEvent *labelEventIterator
+}
+
+// NewIterator create a new iterator
+func NewIterator(projectID, token string, since time.Time) *iterator {
+	return &iterator{
+		gc:       buildClient(token),
+		project:  projectID,
+		since:    since,
+		capacity: 20,
+		issue: &issueIterator{
+			index: -1,
+			page:  1,
+		},
+		note: &noteIterator{
+			index: -1,
+			page:  1,
+		},
+		labelEvent: &labelEventIterator{
+			index: -1,
+			page:  1,
+		},
+	}
+}
+
+// Error return last encountered error
+func (i *iterator) Error() error {
+	return i.err
+}
+
+func (i *iterator) getNextIssues() bool {
+	sort := "asc"
+	scope := "all"
+	issues, _, err := i.gc.Issues.ListProjectIssues(
+		i.project,
+		&gitlab.ListProjectIssuesOptions{
+			ListOptions: gitlab.ListOptions{
+				Page:    i.issue.page,
+				PerPage: i.capacity,
+			},
+			Scope:        &scope,
+			UpdatedAfter: &i.since,
+			Sort:         &sort,
+		},
+	)
+
+	if err != nil {
+		i.err = err
+		return false
+	}
+
+	// if repository doesn't have any issues
+	if len(issues) == 0 {
+		return false
+	}
+
+	i.issue.cache = issues
+	i.issue.index = 0
+	i.issue.page++
+	i.note.index = -1
+	i.note.cache = nil
+
+	return true
+}
+
+func (i *iterator) NextIssue() bool {
+	// first query
+	if i.issue.cache == nil {
+		return i.getNextIssues()
+	}
+
+	if i.err != nil {
```

Comment:
I understand that it won't change anything with the way you implemented it, but it's a good practice to check if there is a sticky error first thing, and do exactly nothing if there is one.